### ISSUE TITLE
Add version to block header

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1135]
We have discussed adding a TendermintVersion and AppVersion to the block header. 

Perhaps we really only need one `Version string` field. 

Additionally, we discussed prefixing the serialized block with a single byte representing the version.

We should discuss the actual requirements and needs here and settle on something.

**Comment by adrianbrink** [comment]
@rigelrozanski @sunnya97 and I just talked about it.

Putting the version number in the block makes a lot of sense, since it forces all validators to upgrade the software in a timely manner. Moreover it makes every change a hardfork which is a good thing.

@jessysaurusrex and I also talked about this as a way to incentivise validators to run the latest software.

**Comment by rigelrozanski** [comment]
@sunnya97 also suggested that we include a... hash or piece of the hash in order to eliminate certain problems associated with hard forks? (what if you have two forks with same version and different hashes right)

**Comment by sunnya97** [comment]
Yeah, what my suggestion was is that the chain id should include somehow the Tendermint version and app version like @ebuchman suggested, but also have it somehow incorporate the like hash of the genesis block or something.  (Btw, I define the genesis block as first block in chain that doesn't have a justification from the previous block).  That way even if the software versions don't change, but you need to fork the validator set, it still gets a new chain id.

**Comment by b00f** [comment]
Can we merge this issue with #1855 ?

**Comment by ebuchman** [comment]
Don't think so - we want to be explicit about this being version numbers.

**Comment by b00f** [comment]
@ebuchman Let me know if you haven't assigned it to anyone. I did same for app_data here: https://github.com/tendermint/tendermint/pull/1953

**Comment by b00f** [comment]
`TendermitVersion` and `AppVersion` of proposer?
